KAFKA-2720: expire group metadata when all offsets have expired

hachikuji opened the pull request.

hachikuji committed:
KAFKA-2720 [WIP]: expire group metadata when all offsets have expired

hachikuji commented:
cc @onurkaraman @guozhangwang 

This patch moves group metadata removal to the periodic offset expiration thread so that group metadata (including the generation) is preserved until all offsets for the group have expired. Since the last member may leave the group before offsets have expired, this patch changes the behavior to increment the generation and essentially write an otherwise empty group metadata object to the log. The main idea is to avoid resetting the generation to 0 until it is safe to do so. This also helps users who currently have no easy way to tell if the group simply doesn't exist or it is dead with possibly some offset state left behind. 

Couple notes on the current patch:
1. I've added a new `Empty` state to the coordinator's group state machine. This may be unnecessary, since it seems `Stable` could work as well, but I preferred slightly to keep the logic for this case separate. It also gets around the need to have a direct transition from `PreparingRebalance` to `Stable`, which should not be possible normally. Instead we allow `PreparingRebalance` to transition to `Empty` when there are no members left in the group.
2. I've done a little bit of additional cleanup. The most notable change is the removal of the offset expiration lock in `GroupMetadataManager`, which seemed unnecessary with a small change to the offset loading logic. Rather than loading offsets directly into the cache, I first stage them in a local collection as is done with the group metadata. This should prevent the expiration thread from seeing stale values, but let me know if I've missed something.

ijuma commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -301,8 +292,9 @@ class GroupCoordinator(val brokerId: Int,
-    delayedGroupStore.foreach(groupManager.store)
+    // for deadlock if the callback is invoked holding other locks (e.g. the replica
+    // state change lock)
+    delayedGroupStore.map(groupManager.store)
```
Why did you change from `foreach` to `map`? It doesn't seem like the return value is being used.

hachikuji replied:
Yeah, you're right. I briefly had considered refactoring the delayed store api and this may have been a side effect of manually reverting.

guozhangwang commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -371,74 +373,84 @@ class GroupMetadataManager(val brokerId: Int,
-                    s"loading partition ${topicPartition}")
-              }
+            loadedOffsets.foreach { case (groupTopicPartition, offsetAndMetadata) =>
+              offsetsCache.put(groupTopicPartition, offsetAndMetadata.copy (
```
Would it possible that another thread is expiring the cache at the same time still?

guozhangwang commented:
Just browsed through the code change, regarding 2) I am still not clear how it prevent concurrent access on the `offsetCache`. Let's discuss offline.

guozhangwang commented:
Specifically, could this issue be likely to triggered even without log compaction stopping? (second incident of the following blog post):

https://engineering.linkedin.com/blog/2016/05/kafkaesque-days-at-linkedin--part-1

hachikuji commented:
Summary of brief offline discussion with @guozhangwang: I think we agree that concurrent loading and expiration is safe with this change, but we are trying to find a way to also solve the problem of expiration with a concurrent offset commit which is unprotected with and without this patch.

onurkaraman commented:
hey @hachikuji I'll try to get to this patch after I finish up testing the topic indexing idea you mentioned in the "[DISCUSS] scalability limits in the coordinator" thread.

guozhangwang commented:
Also ping @jjkoshy for reviews as well since you fixed the problem of race condition between offset expiration and added the lock; we are trying to safely remove the lock, and at the same time fix the race condition between offset expiration and offset commit as well.

Ishiihara commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -83,13 +83,28 @@ private[coordinator] case object Stable extends GroupState { val state: Byte = 3
+  * transition: last offsets removed in periodic expiration task => Dead
+  *             join group from a new member => PreparingRebalance
+  */
+private[coordinator] case object Empty extends GroupState { val state: Byte = 5 }
```
We should also document state transitions from other states to EMPTY state.

Ishiihara commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -645,42 +636,49 @@ class GroupCoordinator(val brokerId: Int,
+          info(s"Group ${group.groupId} with generation ${group.generationId} is now empty")
+          delayedStore = Some(groupManager.prepareStoreGroup(group, Map.empty, errorCode => {
+            if (errorCode != Errors.NONE.code) {
+              // we failed to persist the empty group. if we don't retry (which is how
```
I would like to know how the retry is facilitated here.  Do we do it by signaling the client with an error code of a retriable exception. 

Seems that this will the result of store the group metadata is returned to client. I am not sure that whether this would be the response of JoinGroupRequest or LeaveGroupRequest.

hachikuji replied:
I left a comment here because I'm not sure yet how we should handle it. When we fail to write the state for a generation normally, we transition back to `PreparingRebalance` and force the members to rejoin. This results in sort of a "ghost generation" which never gets written to the log. If we then have a partition emigration before the rebalance completes, then the last successfully written generation will come back to life. I'm not sure if this is a problem, but it seems it may be safe since members never see the rebalance complete until it is successfully written. 

By doing nothing here, we have similar behavior. If partition emigration happens before the group is cleaned up, then the last stable generation will reemerge and the coordinator will probably just time out the members (since they should be gone by now). Again, I am not sure this is actually a problem. If one of those members was still around (maybe because of a temporary network partition), it should be safe for it to continue doing work since no new non-empty generations could have successfully completed a rebalance without writing the new generation to the log.

onurkaraman replied:
It seems that the only reason for persisting group state on transition to Empty is to persist the fact that the generation id incremented. Is that right?

hachikuji replied:
Yes, that's right. I think it also makes it a little easier for users to know when the group is dead (that is, has no members), but nevertheless still has offsets left around. In that case, `DescribeGroup` will return `Empty` instead of `Dead`.

ijuma replied:
The code was merged with this `if` block with no code inside (which I find very confusing) and the comment which is a bit unclear IMO. Can we please fix this in a follow-up? cc @hachikuji @guozhangwang

Ishiihara commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -755,3 +753,13 @@ object GroupCoordinator {
+                           generationId: Int,
+                           subProtocol: String,
+                           leaderId: String,
+                           errorCode: Short)
```
nit: missing empty line.

Ishiihara commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -371,74 +373,84 @@ class GroupMetadataManager(val brokerId: Int,
-                      loadedGroups.remove(groupId)
-                      removedGroups.add(groupId)
-                    }
+                    // special handling for version 0:
```
The comment does not match the implementation.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -83,13 +83,28 @@ private[coordinator] case object Stable extends GroupState { val state: Byte = 3
 private object GroupMetadata {
   private val validPreviousStates: Map[GroupState, Set[GroupState]] =
-    Map(Dead -> Set(Stable, PreparingRebalance, AwaitingSync),
+    Map(Dead -> Set(Stable, PreparingRebalance, AwaitingSync, Empty, Dead),
```
I don't understand why the Dead -> Dead state transition was added. Keeping the allowed transitions to a minimum would go a long way in keeping things easy to understand.

onurkaraman replied:
Intuitively it seems like transitions to Dead should be extremely restricted: just Empty -> Dead. But the implementation now allows every state to transition into the Dead state. Why? Is this because of the group cleanup during handleGroupEmigration?

hachikuji replied:
Yes, that was the case I was thinking about. The group can be in any state when that happens, right? I'm not sure about the Dead -> Dead transition. It seemed nothing protected a concurrent expiration and emigration from both forcing a transition to Dead, but maybe it's actually safe since they are both only executed from the scheduler, which has just one thread.

guozhangwang replied:
Upon group migration, the group is technically "not exist" any more for this coordinator and hence we do not need to care about its status at all. Could we just cleanup the group registry from the cache directly?

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -169,7 +184,7 @@ private[coordinator] class GroupMetadata(val groupId: String, val protocolType:
   def generateMemberIdSuffix = UUID.randomUUID().toString
 
-  def canRebalance = state == Stable || state == AwaitingSync
+  def canRebalance = state == Stable || state == AwaitingSync || state == Empty
```
can be rewritten as `GroupMetadata.validPreviousStates(PreparingRebalance).contains(state)`

hachikuji committed:
move offset cache into GroupMetadata and protect access with the group lock

hachikuji commented:
@onurkaraman @guozhangwang Ready for another look. The main change I made is moving the offset cache of each group into `GroupMetadata`. For groups which only use offset commits, the group will stay in the `Empty` state until all offsets for the group expire. The change helps to unify expiration logic and addresses the race condition between offset commits and offset expiration, which could previously result in an inconsistent cache. Now offset update/expiration is protected with the group lock.

guozhangwang commented:
Thanks @hachikuji , will review.

The Jenkins failure looks familiar to me. I think we already have a JIRA for it?

hachikuji commented:
@guozhangwang Yeah, looks like an instance of [KAFKA-3155](https://issues.apache.org/jira/browse/KAFKA-3155).

guozhangwang commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -39,89 +38,97 @@ import kafka.metrics.KafkaMetricsGroup
   )
 
-  def currentGroups(): Iterable[GroupMetadata] = groupsCache.values
+  def start(enableExpiration: Boolean = true) {
```
This function name is a bit awkward to me:

1) it is only triggered in `GroupCoordinator.startup`, and enableExpiration is always passed (i.e. default value is never used?).

2) it only take effects if the passed `enableExpiration` is true.

Could we rename it to `startMetadataCleanupInBackground` with no parameters, and let coordinator to call it only if `enableExpiration` is true?

hachikuji replied:
Ack. I changed the name to `enableMetadataExpiration`.

guozhangwang commented:
The updated state machine diagram looks reasonable to me, and I will let @onurkaraman to take another check on it. One general comment about naming of `removeGroup` and `removeGroupForPartition`, and `onGroupLoaded`, `onGroupUnloaded`, could we differentiate them better by name the ones, for example: `deleteGroup`, `evictGroupsFromCacheForPartition`, and `onGroupLoadedToCache` `onGroupEvictedFromCache`. And also add some comments on when they could be triggered?

hachikuji committed:
Add isolated tests for GroupMetadataManager

hachikuji committed:
NumOffsets gauge needs synchronized access to group state

hachikuji commented:
@guozhangwang @ijuma @Ishiihara @onurkaraman: I'm removing the WIP tag on this patch. I've cleaned up a few small problems and added additional test cases. Take a look when you have a chance.

guozhangwang commented:
@hachikuji All my comments are from in the previous pass and I do not have further ones, would be OK to go with it if others have made their own passes.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -39,89 +38,98 @@ import kafka.metrics.KafkaMetricsGroup
   newGauge("NumOffsets",
     new Gauge[Int] {
-      def value = offsetsCache.size
+      def value = groupMetadataCache.values.map(group => {
```
This worries me a bit. Every time we compute this metric, we need to hold onto many locks.

hachikuji replied:
Bugs me as well. I debated whether it was worthwhile keeping a separate atomic counter to track the size of the offsets or changing to a concurrent map, but it seemed like premature optimization. If you have any ideas, I'd love to hear them.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -120,10 +141,14 @@ case class GroupSummary(state: String,
+  private val offsets = new mutable.HashMap[TopicPartition, OffsetAndMetadata]
+  private val pendingOffsetCommits = new mutable.HashMap[TopicPartition, OffsetAndMetadata]
+
+  var protocolType: Option[String] = None
```
while the `protocols` may change, `protocolType` shouldn't really change. So It's surprising that this is a var. Is this for groups migrating to new consumer or something?

hachikuji replied:
Yes, that was the main case I had in mind. One alternative might be to set the protocol type as "consumer" by default, but at the moment, the protocol type definition is maintained only on the client. It's also possible prior to this patch to change the protocol type of a group by shutting down all the members and restarting them with a different protocol type. I'm not sure there's a strong reason to preserve this, but I thought we may as well.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadata.scala:
```diff
@@ -206,9 +236,15 @@ private[coordinator] class GroupMetadata(val groupId: String, val protocolType:
+    } else {
+      generationId += 1
+      protocol = null
+      transitionTo(Empty)
```
so now there's an Empty state, a GroupMetadata.isEmpty, and a GroupMetadata.is(Empty). Does it make sense to get rid of isEmpty?

hachikuji replied:
I think it probably does.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -313,34 +337,45 @@ class GroupMetadataManager(val brokerId: Int,
     }
 
+    val group = groupMetadataCache.get(groupId)
+    group.prepareOffsetCommit(offsetMetadata)
```
wouldn't `group.prepareOffsetCommit` need to be in a synchronized block? all other group metadata reads/writes are guarded by a synchronized block.

hachikuji replied:
Good catch.

ijuma commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -39,89 +38,98 @@ import kafka.metrics.KafkaMetricsGroup
    */
   def getGroup(groupId: String): GroupMetadata = {
-      groupsCache.get(groupId)
+      groupMetadataCache.get(groupId)
```
Is there a reason why we don't return `Option[GroupMetadata]` here? Also, indenting should to be fixed.

hachikuji replied:
Using `Option` works for me.

hachikuji committed:
minor refactor/fixes from review comments

hachikuji committed:
Rename start to enableMetadataExpiration in GroupMetadataManager

hachikuji commented:
FYI: There are some system test failures that I'm looking into.

hachikuji committed:
Initialize group in the right state

hachikuji commented:
Found the problem. Here's a nearly clean run: http://testing.confluent.io/confluent-kafka-branch-builder-system-test-results/?prefix=2016-06-15--001.1465997374--hachikuji--KAFKA-2720--9ffe4d0/. The two failures appear to be unrelated.

guozhangwang commented:
LGTM. Merged to trunk.

ijuma commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -453,18 +478,45 @@ class GroupMetadataManager(val brokerId: Int,
    * that partition.
-   * @param offsetsPartition Groups belonging to this partition of the offsets topic will be deleted from the cache.
+    *
+    * @param offsetsPartition Groups belonging to this partition of the offsets topic will be deleted from the cache.
```
Looks like an unintentional change, fix it in a follow-up? cc @hachikuji

ijuma commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -475,31 +527,22 @@ class GroupMetadataManager(val brokerId: Int,
-        // clear the offsets for this partition in the cache
-
         /**
          * NOTE: we need to put this in the loading partition lock as well to prevent race condition of the leader-is-local check
```
This is no longer using the `loadingPartitions` lock.

hachikuji commented:
Thanks @ijuma. I'll prepare a follow-up.
